src/Build/Collections/RetrievableEntryHashSet/HashSet.cs

```diff
@@ -592,8 +574,8 @@ public virtual void OnDeserialization(Object sender)
             }
 
             int capacity = _siInfo.GetInt32(CapacityName);
-            _comparer = (IEqualityComparer<IKeyed>)_siInfo.GetValue(ComparerName, typeof(IEqualityComparer<IKeyed>));
-            _constrainedComparer = _comparer as IConstrainedEqualityComparer<IKeyed>;
+            _comparer = (IEqualityComparer<string>)_siInfo.GetValue(ComparerName, typeof(IEqualityComparer<string>));
```

Comment:
Note these types are serializable - so this breaks the serialization format, we do we expect this to be round tripped with other MSBuild versions?

Reply:
Shouldn't be. There are a few instances where caches are created and read in between builds so the format is important. I think I've added a version field to those now to invalidate them, so shouldn't be a big deal. But I don't think this is used in any of those (I will double check). For the main usage of binary serialization there's a handshake takes place that will ensure it's running exactly the same version of MSBuild on both ends of the pipe.